Extensions/xkit_patches.js

```diff
@@ -915,6 +915,208 @@ XKit.extensions.xkit_patches = new Object({
+				 * Create a specification for a control button that can be added to
+				 * future posts using `XKit.interface.add_control_button`.
+				 * @param {String} class_name - CSS class of the button to be created
+				 * @param {String} icon - URL of the button's icon
+				 * @param {String} text - Hover text of the button
+				 * @param {EventListener} func - Function called on click of control button
+				 * @param {String?} ok_icon - URL of icon displayed when the button is
+				 *                            "completed" (e.g. reblog button turning green)
+				 */
+				create_control_button: async function(class_name, icon, text, func, ok_icon) {
+					if (this.control_button_template == null) {
+						this.control_button_template = await this.get_control_button_template();
+					}
+
+					XKit.interface.added_icon.push(class_name);
+					XKit.interface.added_icon_icon.push(icon);
+					XKit.interface.added_icon_text.push(text);
+
+					XKit.tools.add_css(`.${class_name} .xkit-interface-icon {
+						background-image: url('${icon}');
+						background-size: 100% 100%;
+						width: 21px;
+						height: 21px;
+					}`, `xkit_interface_icon__${class_name}`);
+
+					if (typeof ok_icon !== "undefined") {
+						XKit.tools.add_css(`.${class_name} .xkit-interface-icon-completed {
+							background-image: url('${ok_icon}');
+						}`, `xkit_interface_icon__completed__${class_name}`);
+					}
+
+					$(document).on('click', '.' + class_name, function() {
+						if ($(this).hasClass("xkit-interface-working") || $(this).hasClass("xkit-interface-disabled")) { return; }
+						if (typeof func === "function") { func.call(this, event); }
+					});
+				},
+				add_control_button: async function(obj, class_name, additional) {
+					if (typeof additional == "undefined") {additional = ""; }
+
+					if (XKit.interface.added_icon.indexOf(class_name) === -1) {
+						return;
+					}
+
+					var m_text = XKit.interface.added_icon_text[XKit.interface.added_icon.indexOf(class_name)];
+
+					var post_obj = await XKit.interface.react.post(obj);
+					var post_id = post_obj.id;
+					var post_type = post_obj.type;
+					var post_permalink = post_obj.permalink;
+
+					var m_data = `data-post-id = "${post_id}" data-post-type="${post_type}" data-permalink="${post_permalink}"`;
+
+					var m_html = this.control_button_template
+						.replace(/{{className}}/g, class_name)
+						.replace(/{{text}}/g, m_text)
+						.replace(/{{additional}}/g, additional)
+						.replace(/{{data}}/g, m_data)
+					;
+
+					// we know that XKit.css_map.getCssMap() has been called because we have a template from create_control_button
+					// so we skip that call with this XKit.css_map.keyToCss() call.
+					var controlsSelector = XKit.css_map.keyToCss("controls");
+					var controls = $(obj).find(controlsSelector);
+
+					if (controls.length > 0) {
+						controls.prepend(m_html);
+					}
+
+					if (XKit.interface.where().search) {
+						XKit.interface.trigger_reflow();
+					}
```

Comment:
`interface.react` shouldn't call any regular `interface` methods since they will all be broken on react

```suggestion
```